modules/common/src/timeout.ts

```diff
@@ -46,9 +46,21 @@ export function promiseWithTimeout<T>(
   return new Promise((resolve, reject) => {
     setTimeout(
       () => reject(new TimeoutError(timeoutMessage || "Time out")),
-      timeout
+      timeout,
     );
 
-    return prom.then(result => resolve(result)).catch(err => reject(err));
+    return prom.then((result) => resolve(result)).catch((err) => reject(err));
+  });
+}
+
+export async function functionRetryWithCondition<T>(
+  func: () => Promise<T>,
+  // should not throw to execute recursivity
+  retryCondition: (e: Error) => Promise<void>,
+): Promise<T> {
+  return func().catch((e) => {
+    return retryCondition(e).then(() =>
+      functionRetryWithCondition(func, retryCondition)
+    );
```

Comment:
Wouldn't it be easier to have this?

```ts
async function tryUntil({
  func,
  onFail,
  shouldContinue
}) {
  do {
    try {
      const result = await func();
      // it worked
      return result;
    }
    catch (err) {
      onFail();
    }
  } while(shouldContinue())
}

// Usage:

await tryUntil({
  func: async () => {
    const result = await this.graphql.request(...);
    // ...
    return result;
  },
  onFail: () => {
    retries--;
    this.graphql.shiftClient()
  },
  shouldContinue: () => {
    return retries > 0;
  }
});
```

But actually, this logic could be abstracted inside of `request` and not need to be copied in every single method, every time.